rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+and these services use the splitting API as a part of their implementation.
+
+#### Master API
+
+The master is responsible for registering datasets, generating and tracking
+iteration and worker ids, and generating dataset splits for processing on
+workers.
+
+Below is a sketch of the Master API. This API is not public and is subject to
+change.
+
+```cpp
+/// ---- Methods called by consumers ----
+
+// Registers a dataset and returns an id for the dataset. If the dataset is
+// already registered, its dataset id is returned.
+int GetOrRegisterDataset(GraphDef dataset);
+
+// Creates and returns `num_consumers` iterator ids which partition the
+// specified dataset. This also creates an internal `iteration_id` used to
+// track the overall dataset iteration. `num_tasks` defines how many tasks to
+// create. If `num_tasks` is -1, it is up to the master to determine how many
+// tasks to create.
+list<int> CreateIterators(int dataset_id, int num_consumers,
+                          int num_tasks);
+
+// Returns the list of tasks processing data for `iterator_id`. Consumers query
+// this to find which worker addresses to read data from.
+list<TaskInfo> GetWorkersForiterator(int iterator_id);
+
+///---- Methods called by input workers ----
+
+// Registers a worker and returns its worker id.
+int RegisterWorker(WorkerInfo worker_info);
+
+// Requests the next splits to process on the given worker for the given
+// iteration_id.
+List<Split> GetSplits(int worker_id, int iteration_id);
+```
+
+#### Worker API
+
+The worker is responsible for processing datasets and providing dataset elements
+to consumers.
+
+Below is a sketch of the Worker API. This API is not public and is subject to
+change.
+
+```cpp
+/// ---- Methods called by consumers ----
+
+// Gets the next element for the specified iterator_id.
+list<Tensors> GetElement(iterator_id);
+
+/// ---- Methods called by master ----
+
+// Requests that the worker process the specified dataset. This will trigger the
+// worker to start requesting splits from the master using the `iteration_id`.
+void ProcessDataset(int dataset_id, int iteration_id, list<int> iterator_ids);
+```
+
+#### Visitation Guarantees
+
+When iterating over a dataset, the tf.data service will process all input data
+at least once, even in the presence of master or worker failures. If there are
+no failures, all input data will be processed exactly once.
+
+With determinstic execution enabled, the tf.data service provides an
+exactly-once visitation guarantee even in the face of master or worker failures.
+
+#### Determinism
+
+Deterministic processing is a cornerstone of tf.data. Determinism is valuable
+for debugging and experimentation. This section discusses how the tf.data
+service will provide determinism.
+
+To get deterministic behavior, the tf.data service will require three things:
+
+1.  The dataset being distributed has deterministic output.
+1.  The user sets `deterministic=True` when calling
```

Comment:
Why requiring determinism to be opt-in instead of making non-determinism opt-in like the rest of the tf.data API?

Reply:
I ask because I can think of some not-too-expensive ways of enforcing exactly-once, like if consumers request a <split_id, batch_within_split> instead of just "next batch please" in the RPC, and things like that.

Reply:
We want to give good out of the box performance for users using the tf.data service to speed up input processing. Achieving deterministic behavior is more expensive in a distributed environment than in a single-host environment because we need to use a fixed number of tasks, we can't dynamically balance load, and we can't sloppily interleave results. We expect the common case to be a desire for performance over determinism.

Let me think a little bit more about how to achieve exactly-once visitation with non-deterministic service execution. I'll get back to you

Reply:
I've updated the proposal to support an exactly-once guarantee even when tf.data service processing is non-deterministic.